ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -0,0 +1,256 @@
+                  "is not the same as the correlation id in the associated GetRequest: " + correlationId,
+                  RouterErrorCode.UnexpectedInternalError));
+          operationTracker.onResponse(getRequestInfo.replicaId, false);
+          // we do not notify the ResponseHandler responsible for failure detection as this is an unexpected error.
+        } else {
+          processGetResponse(getRequestInfo, getResponse);
+        }
+      } catch (IOException | MessageFormatException e) {
+        // This should really not happen. Again, we do not notify the ResponseHandler responsible for failure
+        // detection.
+        setOperationException(new RouterException("Response deserialization received an unexpected error", e,
+            RouterErrorCode.UnexpectedInternalError));
+        operationTracker.onResponse(getRequestInfo.replicaId, false);
+      }
+    }
+    checkAndMaybeComplete();
+  }
+
+  /**
+   * Process the actual GetResponse extracted from a {@link ResponseInfo}
+   * @param getRequestInfo the associated {@link RequestInfo} for which this response was received.
+   * @param getResponse the {@link GetResponse} extracted the {@link ResponseInfo}
+   * @throws IOException if there is an error during deserialization of the GetResponse.
+   * @throws MessageFormatException if there is an error during deserialization of the GetResponse.
+   */
+  void processGetResponse(GetRequestInfo getRequestInfo, GetResponse getResponse)
+      throws IOException, MessageFormatException {
+    ServerErrorCode getError = getResponse.getError();
+    if (getError == ServerErrorCode.No_Error) {
+      int partitionsInResponse = getResponse.getPartitionResponseInfoList().size();
+      // Each get request issued by the router is for a single blob.
+      if (partitionsInResponse != 1) {
+        setOperationException(new RouterException("Unexpected number of partition responses, expected: 1, " +
+            "received: " + partitionsInResponse, RouterErrorCode.UnexpectedInternalError));
+        operationTracker.onResponse(getRequestInfo.replicaId, false);
+        // Again, no need to notify the responseHandler.
+      } else {
+        getError = getResponse.getPartitionResponseInfoList().get(0).getErrorCode();
+        responseHandler.onRequestResponseError(getRequestInfo.replicaId, getError);
+        if (getError == ServerErrorCode.No_Error) {
+          handleBody(getResponse.getInputStream());
+          operationTracker.onResponse(getRequestInfo.replicaId, true);
+        } else {
+          // process and set the most relevant exception.
+          processServerError(getError);
+          if (getError == ServerErrorCode.Blob_Deleted || getError == ServerErrorCode.Blob_Expired) {
+            // this is a successful response and one that completes the operation regardless of whether the
+            // success target has been reached or not.
+            operationCompleted = true;
```

Comment:
Anyway the operation is completed not based on the operation tracker, but not hurt to update its status for completeness. I.e., the operation tracker still records from how many replicas it have receive responses.